lib/streamlit/elements/widgets/multiselect.py

```diff
@@ -291,76 +262,72 @@ def _multiselect(
-        check_cache_replay_rules()
-        check_callback_rules(self.dg, on_change)
-        check_session_state_rules(default_value=default, key=key)
-        maybe_raise_label_warnings(label, label_visibility)
+        check_widget_policies(self.dg, key, on_change, default)
 
-        opt = ensure_indexable(options)
-        check_python_comparable(opt)
+        widget_name = "multiselect"
+        maybe_raise_label_warnings(label, label_visibility)
 
-        indices = _check_and_convert_to_indices(opt, default)
+        indexable_options = ensure_indexable_and_comparable(options)
+        formatted_options = [format_func(option) for option in indexable_options]
+        default_values = get_default_indices(indexable_options, default)
 
-        id = compute_widget_id(
-            "multiselect",
+        form_id = current_form_id(self.dg)
+        widget_id = compute_widget_id(
+            widget_name,
             user_key=key,
             label=label,
-            options=[str(format_func(option)) for option in opt],
-            default=indices,
+            options=formatted_options,
+            default=default_values,
             key=key,
             help=help,
             max_selections=max_selections,
             placeholder=placeholder,
-            form_id=current_form_id(self.dg),
+            form_id=form_id,
             page=ctx.active_script_hash if ctx else None,
         )
 
-        default_value: list[int] = [] if indices is None else indices
-
-        multiselect_proto = MultiSelectProto()
-        multiselect_proto.id = id
-        multiselect_proto.label = label
-        multiselect_proto.default[:] = default_value
-        multiselect_proto.options[:] = [str(format_func(option)) for option in opt]
-        multiselect_proto.form_id = current_form_id(self.dg)
-        multiselect_proto.max_selections = max_selections or 0
-        multiselect_proto.placeholder = placeholder
-        multiselect_proto.disabled = disabled
-        multiselect_proto.label_visibility.value = get_label_visibility_proto_value(
+        proto = MultiSelectProto()
+        proto.id = widget_id
+        proto.default[:] = default_values
+        proto.form_id = form_id
+        proto.disabled = disabled
+        proto.label = label
+        proto.max_selections = max_selections or 0
+        proto.placeholder = placeholder
+        proto.label_visibility.value = get_label_visibility_proto_value(
             label_visibility
         )
-
+        proto.options[:] = formatted_options
         if help is not None:
-            multiselect_proto.help = dedent(help)
-
-        serde = MultiSelectSerde(opt, default_value)
+            proto.help = dedent(help)
 
+        serde = MultiSelectSerde(indexable_options, default_values)
         widget_state = register_widget(
             "multiselect",
-            multiselect_proto,
-            user_key=key,
+            proto,
+            # user_key=key,
```

Comment:
?

Reply:
The `user_key` parameter is not used by the `register_widget` function anymore for some time as it looks like, so I commented it out to highlight this 🙁 A better way would be to refactor the `register_widget` function and clean this up for the entire code; I will file a tech debt item and comment this in again.